src/app/components/services/chart.service.js

```diff
@@ -0,0 +1,710 @@
+          label.precision = label.precision ? Math.min(label.precision, precision) : precision;
+        }
+
+        return label.ids;
+      };
+
+      /**
+       * Remove all the filters applied to the label dimensions
+       * TODO: Add the possibility to reapply removed filters
+       * @private
+       */
+      var _removeAllLabelFilters = function () {
+        _.each(chartService.cs.dims, function (dim) {
+          dim.filterAll();
+        });
+        dc.filterAll();
+        dc.redrawAll();
+      };
+
+      /**
+       * Create the Crossfilter instance from the subject data
+       * @private
+       */
+      var _populateCohortCrossfilter = function () {
+        _removeAllLabelFilters();
+        chartService.cs.cross.remove();
+        chartService.cs.cross.add(chartService.cs.subjects);
+      };
+
+      /**
+       * Fetch the data for the selected node
+       * @param node
+       * @returns {*}
+       */
+      chartService.addNodeToActiveCohortSelection = function (node, filters) {
+
+        var _filter, _deferred = $q.defer();
+
+        var _getFilter = function (label, filters) {
+          return _.find(filters, {label: label});
+        };
+
+        // Get all observations under the selected concept
+        node.restObj.one('observations').get().then(function (observations) {
+          observations = observations._embedded.observations;
+
+          observations.forEach(function (obs) {
+            if (obs.value !== null) {
+
+              if (filters) {
+                _filter = _getFilter(obs.label, filters);
+              }
+
+              // Add the concept to the list of chart labels
+              var _id = _addLabel(obs, node, _filter);
+
+              // Check if the subject of the observation is already present
+              var found = _.find(chartService.cs.subjects, {id: obs._embedded.subject.id});
+
+              if (found) {
+                found.labels[_id] = obs.value;
+              } else {
+                obs._embedded.subject.labels = {};
+                obs._embedded.subject.labels[_id] = obs.value;
+                chartService.cs.subjects.push(obs._embedded.subject);
+              }
+            }
+          });
+
+          // Add all the subjects to a crossfilter instance
+          _saveFilters();
+
+          _populateCohortCrossfilter();
+
+          // Notify the applicable controller that the chart directive instances
+          // can be created
+          //$rootScope.$broadcast('prepareChartContainers', chartService.cs.labels);
+          GridsterService.resize('#main-chart-container', chartService.cs.labels, false);
+
+          _reapplyFilters();
+
+          _deferred.resolve();
+
+        }, function (err) {
+          //TODO: add alert
+          _deferred.reject('Cannot get data from the end-point.' + err);
+        });
+
+        return _deferred.promise;
+      };
+
+      /**
+       *
+       */
+      chartService.onNodeDrop = function (node) {
```

Comment:
Not sure if it is good practice to put event handlers in a service instead of a controller, even if it is called indirectly, but I guess we need to refactor this anyway.